### PR TITLE
Site Export: ensure that the export endpoint uses Gutenberg theme classes

### PR DESCRIPTION
**Pull request by ramonjd**
## What?
Extending the `WP_REST_Edit_Site_Export_Controller` endpoint to use Gutenberg theme methods.

> [!NOTE]
> This PR doesn't require backporting - the code is taken from existing Core methods/functions from 6.0/5.9. The references to Core classes and functions have been changed to `_Gutenberg/gutenberg_` only.

## Why?

Like it says in the title, ensure that the export endpoint uses Gutenberg theme classes.

It makes sure that theme exports get the benefit of the latest changes to theme json and resolver.

For example, `style.background` is a new theme.json style property, but it won't be exported by the Core endpoint until version 6.6.

This endpoint hasn't had any backwards compat extensions in Gutenberg for the last few releases and probably further back so it's probably worth doing.

## How?
Copying over the public methods and swapping WP functions/methods with Gutenberg ones 😄 

The changes live in the `/lib` folder because none of the changes are backwards compat specific. Props to @andrewserong for pointing this out 😄 

Having this extension permanently in Gutenberg means that theme.json exporting will always use the latest version of the Theme JSON family of classes. 

## Testing Instructions

Add a background image to a site in the site editor. 

Screenshot 2024-05-10 at 5 18 15 pm

Save the site

Export your theme

Screenshot 2024-05-10 at 5 18 23 pm

Check the exported theme.json - your background styles should be featured in `style.background`.

Example snippet from my exported theme.json:

```json
	"styles": {
		"background": {
			"backgroundImage": {
				"id": 26,
				"source": "file",
				"title": "pizza1",
				"url": "http://localhost:8888/wp-content/uploads/2024/05/pizza1.jpg"
			}
		},
		"color": {
			"text": "#333"
		}
	},
```

### CODE REVIEW

File: lib/compat/wordpress-6.6/block-template-utils.php [hunk 1]
```diff
@@ -360,3 +360,111 @@ function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_t
...
+ * specified path in a ZIP file.
+ *
+ * @since 5.9.0
+ * @since 6.0.0 Adds the whole theme to the export archive.
+ *
+ * @global string $wp_version The WordPress version string.
+ *
+ * @return WP_Error|string Path of the ZIP file or error on failure.
+ */
+function gutenberg_generate_block_templates_export_file() {
```
Comment (ramonjd) [inline lib/compat/wordpress-6.6/block-template-utils.php]: This is all just copy paste and swapping out the WP functions for the Gutenberg juicy goodness.

File: lib/compat/wordpress-6.6/class-gutenberg-rest-edit-site-export-controller-6-6.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+class WP_REST_Edit_Site_Export_Controller_6_6 extends WP_REST_Edit_Site_Export_Controller {
+	/**
+	 * Output a ZIP file with an export of the current templates
+	 * and template parts from the site editor, and close the connection.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return WP_Error|void
+	 */
+	public function export() {
```
Comment (ramonjd) [inline lib/compat/wordpress-6.6/class-gutenberg-rest-edit-site-export-controller-6-6.php]: Also a 1:1 copy but using gutenberg_generate_block_templates_export_file()

File: lib/load.php [hunk 3]
```diff
@@ -46,6 +46,11 @@ function gutenberg_is_experiment_enabled( $name ) {
 	require_once __DIR__ . '/compat/wordpress-6.5/class-gutenberg-rest-global-styles-revisions-controller-6-5.php';
 	require_once __DIR__ . '/compat/wordpress-6.5/rest-api.php';
 
+	// WordPress 6.6 compat.
+	require_once __DIR__ . '/compat/wordpress-6.6/class-gutenberg-rest-edit-site-export-controller-6-6.php';
+	require __DIR__ . '/compat/wordpress-6.6/class-gutenberg-rest-templates-controller-6-6.php';
+	require __DIR__ . '/compat/wordpress-6.6/rest-api.php';
```
Comment (ramonjd) [inline lib/load.php]: Separate PR for this here https://github.com/WordPress/gutenberg/pull/61564 in case this PR languishes

**Comment by ramonjd** [comment]
> Example snippet from my exported theme.json:

Gutenberg preserves the id/source/title props added by [block upload](https://github.com/WordPress/gutenberg/blob/08755b138a1ff20b972e85022ab25f214b137f80/packages/block-editor/src/components/global-styles/background-panel.js#L247-L247).

```json
				"id": 26,
				"source": "file",
				"title": "pizza1",
```

I guess they should be reflected in theme.json at some point.

**Review by andrewserong — COMMENTED** [review]
This is testing well for me, with the output containing the uploaded image and metadata stored via the uploaded version of the image:

```
	"styles": {
		"background": {
			"backgroundImage": {
				"id": 4727,
				"source": "file",
				"title": "neuschwastein-castle-5",
				"url": "http://a-toule-site.local/wp-content/uploads/2024/01/neuschwastein-castle-5.jpg"
			}
		},
```

One thought while looking at the code change — is there anything about this change that's particular to 6.6 or requires backporting? If not, would it be better to include these changes in either `/lib` or `/lib/experimental` to both flag that no backporting is required, and so that the GB versions are still present for subsequent releases? Otherwise, we might wind up needing to re-add the files further down the track when the 6.6 compat directory is eventually removed 🤔

**Comment by ramonjd** [comment]
> One thought while looking at the code change — is there anything about this change that's particular to 6.6 or requires backporting? If not, would it be better to include these changes in either /lib or /lib/experimental to both flag that no backporting is required, and so that the GB versions are still present for subsequent releases? Otherwise, we might wind up needing to re-add the files further down the track when the 6.6 compat directory is eventually removed

Oh yeah, that's a very good point 😄 

I think I was hypnotized by the fact that background images in theme.json is a 6.6 feature, but you're right, this will be an ongoing concern so it should be maybe in lib, next to global styles controller? What do you reckon?

**Comment by andrewserong** [comment]
> so it should be maybe in lib, next to global styles controller

Yep, that sounds good to me. I'd lean toward `lib` if it's identical to `wordpress-develop` except for calling GB versions of things, and `lib/experimental` if it's making changes that aren't present in core.

**Comment by ramonjd** [comment]
> Yep, that sounds good to me. I'd lean toward lib if it's identical to wordpress-develop except for calling GB versions of things, and lib/experimental if it's making changes that aren't present in core.

Done!

**Review by andrewserong — APPROVED** [review]
This is testing nicely for me in a WP 6.5.3 testing environment, with the background image data now included in the export:

image

Also double-checked that the diffs between the included functions and those in `wordpress-develop` match, aside from the changes here that reference Gutenberg versions of classes and functions.

LGTM 🚀

**Comment by ramonjd** [comment]
I appreciate the thorough review @andrewserong 

This is a good one to get in 👍🏻

**Comment by ellatrix** [comment]
Should this be backported for WP 6.6? If so, please add the `Backport to WP Beta/RC` label. Thanks!

**Comment by ramonjd** [comment]
> Should this be backported for WP 6.6? If so, please add the Backport to WP Beta/RC label. Thanks!

This one is just Gutenberg - it's got the label "No Core Sync Required" 👍🏻

**Pull request merged**